### PR TITLE
Support specific version ID declaration for Modrinth

### LINKED ISSUE #49: feat(modrinth): support project references qualified by specific version ID
Supports https://github.com/itzg/docker-minecraft-server/issues/1650

image

### PR DESCRIPTION
**Pull request by JustMangoT**
closes #49 

Using colon to declare specific version id of a Modrinth's mod. 
Also, I'm not sure if this is the best solution for this; I checked #74 and there is a suggestion to use ``mod#version``.

### CODE REVIEW

**Comment by JustMangoT** [comment]
Btw I noticed this may lead to some kind of unexpected errors such as incorrect version ID or incompatible loader, etc... Should those be handled as well?

**Comment by itzg** [comment]
> Btw I noticed this may lead to some kind of unexpected errors such as incorrect version ID or incompatible loader, etc... Should those be handled as well?

It's not the prettiest, but should already throw FailedRequestException with a 404, so I'm thinking don't add any specific logic yet to handle bad IDs. Is that the behavior you've seen?

File: src/main/java/me/itzg/helpers/modrinth/ModrinthCommand.java [hunk 1]
```diff
@@ -236,8 +247,14 @@ private Stream<? extends Path> processProject(String projectRef) {
```
Comment (itzg) [inline src/main/java/me/itzg/helpers/modrinth/ModrinthCommand.java:247]: What you have is really good.
To keep the version splitting local to this method, perhaps you could do the split here with limit 2.

File: src/main/java/me/itzg/helpers/modrinth/ModrinthCommand.java [hunk 2]
```diff
@@ -236,8 +247,14 @@ private Stream<? extends Path> processProject(String projectRef) {
 
         final Project project = getProject(projectRef);
```
Comment (itzg) [inline src/main/java/me/itzg/helpers/modrinth/ModrinthCommand.java:248]: with that you can unconditionally pass parts[0]

File: src/main/java/me/itzg/helpers/modrinth/ModrinthCommand.java [hunk 3]
```diff
@@ -236,8 +247,14 @@ private Stream<? extends Path> processProject(String projectRef) {
...
         if (projectsProcessed.add(project.getId())) {
-            final List<Version> versions = getVersionsForProject(project.getId());
-            final Version version = pickVersion(versions);
+            final Version version;
+            if (projectRef.contains(":")) {
```
Comment (itzg) [inline src/main/java/me/itzg/helpers/modrinth/ModrinthCommand.java:251]: and here conditional on parts.length == 2

**Review by itzg — COMMENTED** [review]
Thanks so much for working on this one. It'll be very useful for people. 

I'm good to merge, but was wondering if you could consider my suggestion?

**Comment by JustMangoT** [comment]
I'm not sure, I just said that because I was thinking of adding another layer to check if the version belongs to the project since the API routes only requires the version ID.

**Comment by itzg** [comment]
Yeah, let's keep it easy for now and just let it throw as is.

File: src/main/java/me/itzg/helpers/modrinth/ModrinthCommand.java (hunk 3 above)
Comment (JustMangoT) [reply src/main/java/me/itzg/helpers/modrinth/ModrinthCommand.java:251]: Should I handle the length > 3 as well or just fallback to the default method?

Comment (itzg) [reply src/main/java/me/itzg/helpers/modrinth/ModrinthCommand.java:251]: The cool thing is if you use [this one](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/String.html#split(java.lang.String,int)) you can limit the split to no more than 2 elements.

**Comment by JustMangoT** [comment]
image
tested

**Review by itzg — APPROVED** [review]
Thanks!

**Comment by itzg** [comment]
Release is ready to be picked up in docker-minecraft-server https://github.com/itzg/mc-image-helper/releases/tag/1.24.19

**Pull request merged**